### PR TITLE
Quilt support

### PR DESCRIPTION
**Pull request by Scrumplex**
Implement support for Quilt Loader and Mappings.

TODO:

- [x] Support Quilt in meta (See https://github.com/PolyMC/meta/pull/6)
- [x] Initial Quilt support
- [x] Select Quilt Mapping version dynamically instead of this hardcoded crap
- [ ] Wait for Quilt Beta release

> NOTE: We only plan to merge this once Quilt is actually released (or is very close to being released)

### CODE REVIEW

**Comment by LennyMcLennington** [comment]
Can someone tell me how quilt is different to Fabric and why it needs to be supported

**Comment by clydebottoms** [comment]
> Can someone tell me how quilt is different to Fabric and why it needs to be supported
Last time I checked, they literally didn't change anything except change every mention of fabric in the code to quilt. Maybe there's a private repo with actual code changes, but I really don't see why quilt needs to exist in the first place. The only reason for the fork seemed to be discord drama, and I don't see what that has to do with the actual fabric mod loader.

**Comment by DioEgizio** [comment]
> > Can someone tell me how quilt is different to Fabric and why it needs to be supported
> 
> Last time I checked, they literally didn't change anything except change every mention of fabric in the code to quilt. Maybe there's a private repo with actual code changes, but I really don't see why quilt needs to exist in the first place. The only reason for the fork seemed to be discord drama, and I don't see what that has to do with the actual fabric mod loader.

bruh no

**Comment by DioEgizio** [comment]
> Can someone tell me how quilt is different to Fabric and why it needs to be supported


https://quilt.gareth-coles.dev/about/faq/#why-should-i-use-quilt

https://quilt.gareth-coles.dev/about/faq/developers/#why-should-i-choose-quilt-for-my-mods-instead-of-fabric

Also quilt is not released yet

**Comment by timoreo22** [comment]
I do think we should be inclusive with the mod loaders we give, still we do have to wait for quilt to release

**Comment by Scrumplex** [comment]
Quilt has a similar spirit as PolyMC. It seems rather natural to support it as well, especially as it's trivially easy to implement

**Comment by clydebottoms** [comment]
> > Can someone tell me how quilt is different to Fabric and why it needs to be supported
> 
> https://quilt.gareth-coles.dev/about/faq/#why-should-i-use-quilt
> 
> https://quilt.gareth-coles.dev/about/faq/developers/#why-should-i-choose-quilt-for-my-mods-instead-of-fabric
> 
> Also quilt is not released yet

Why can't that stuff just be implemented into fabric? Do they not want those changes or what?

**Comment by DioEgizio** [comment]
> Quilt has a similar spirit as PolyMC. It seems rather natural to support it as well, especially as it's trivially easy to implement

I agree



> I do think we should be inclusive with the mod loaders we give, still we do have to wait for quilt to release

Yeah also now it's confirmed, quilt beta on April 20th
IMG_20220226_211318

**Comment by Scrumplex** [comment]
Rebased on develop

**Comment by Scrumplex** [comment]
This now works with modern versions of PolyMC

File: launcher/minecraft/ComponentUpdateTask.cpp
Comment (DioEgizio) [inline launcher/minecraft/ComponentUpdateTask.cpp]: quilt mappings are thought for mappings for mods iirc, what's the point of being there

Comment (Scrumplex) [reply launcher/minecraft/ComponentUpdateTask.cpp]: IIRC they are a drop-in replacement for Fabric's mappings. In earlier versions of this PR I used Quilt mappings instead. But currently meta enforces Fabric Mappings

Comment (DioEgizio) [reply launcher/minecraft/ComponentUpdateTask.cpp]: oh. weird tho

Comment (DioEgizio) [reply launcher/minecraft/ComponentUpdateTask.cpp]: immagine
 seeing upstream pr, they're using hashed here tho (which makes sense since https://github.com/QuiltMC/rfcs/blob/master/rfc/0019-hashed-mojmap.md , but it won't be implemented on initial beta so it'd be fine this way for now)

**Comment by Un1q32** [comment]
I hadn't even heard of quilt before now.  seems cool

**Comment by Scrumplex** [comment]
Ready for review. I would like to merge the meta PR and deploy it, right before merging this one.

File: launcher/ui/pages/instance/VersionPage.cpp
Comment (DioEgizio) [inline launcher/ui/pages/instance/VersionPage.cpp]: technically Quilt could run on any version that intermediary supports

Comment (Scrumplex) [reply launcher/ui/pages/instance/VersionPage.cpp]: I have asked the Quilt guys. This checks out

**Review by DioEgizio — APPROVED** [review]
lgtm personally

File: launcher/ui/pages/instance/ModFolderPage.cpp
Comment (flowln) [inline launcher/ui/pages/instance/ModFolderPage.cpp]: This looks really non-intuitive :/
Can we add a `ModAPI::None` to the enum and use it instead? Doesn't seem like a big change

File: launcher/ui/pages/modplatform/ModPage.cpp
Comment (flowln) [inline launcher/ui/pages/modplatform/ModPage.cpp]: Maybe we could have a function in `ModAPI.h` for returning the string of the mod loaders, since `ModrinthAPI::getModLoaderString` does the same thing basically.

**Review by flowln — COMMENTED** [review]
looks nice, just some nitpicks :p

File: launcher/ui/pages/instance/ModFolderPage.cpp
Comment (Scrumplex) [reply launcher/ui/pages/instance/ModFolderPage.cpp]: Yeah, thought about it as well. Maybe just rename `Any` to `None` or `Unspecified`?

Comment (flowln) [reply launcher/ui/pages/instance/ModFolderPage.cpp]: Maybe `Unspecified` would be nice. Ideally we would have both `Any` and `None`, but since they mean virtually the same thing from the code's perspective, it's fine i think :)

**Review by flowln — APPROVED** [review]
looks good to me :)

**Review by ZekeZDev — APPROVED** [review]

**Review by txtsd — APPROVED** [review]
Quilt is not on Meta yet so I can't test past the `Install Quilt` button, but the code LGTM!

**Review by d-513 — APPROVED** [review]

**Comment by Scrumplex** [comment]
Closes #406

**Pull request merged**
